### ISSUE TITLE
みずしまレビュー

### ISSUE THREAD

**Issue by mizushima1226** [generationokd/raisetech_todo#1]
このissueにみずしまのレビューをぶら下げていきます〜💪

**Comment by mizushima1226** [comment]
トースト系のhooks切り出す感じ、個人的にめっちゃ好きですっ
カスタムフックで抽象化しておくとトーストのライブラリ変更したとしてもカスタムフックだけ修正すれば良いのでメンテナンス性考えるとGoodだと思いました👍

https://github.com/generationokd/raisetech_todo/blob/7c7492b85b7b1f040cc3fa0fca8633866d7ffe9c/src/hooks/useMessage.ts#L1-L26

**Comment by mizushima1226** [comment]
ナイスバリデーションですっ
`duration`、5000で統一してもいいかなぁと思いましたが、5000と2000で使い分けしているんですね💡
どんなこだわりがあるのか気になります👀👀👀
https://github.com/generationokd/raisetech_todo/blob/7c7492b85b7b1f040cc3fa0fca8633866d7ffe9c/src/pages/TodoList.tsx#L50-L57

**Comment by mizushima1226** [comment]
見た感じだとおそらく本当に必要な情報（提供したい情報）は`planList`と`completeList`でしょうか？👀
それだと`apiData`はステートとして保持はせず、axios.getのレスポンスを受け取っているthen句の中で
`planList`と`completeList`にデータを設定するのが良さそうです💡
（実装の意図が伝わりやすくなる気がします💡）

https://github.com/generationokd/raisetech_todo/blob/7c7492b85b7b1f040cc3fa0fca8633866d7ffe9c/src/hooks/useInitialData.ts#L20-L27

**Comment by mizushima1226** [comment]
https://github.com/generationokd/raisetech_todo/blob/7c7492b85b7b1f040cc3fa0fca8633866d7ffe9c/src/hooks/useInitialData.ts#L43-L70

- for文のiの設定、データが多いので10や20区切りで絞り込んでいる感じですかね👀
あくまでも今回は練習で、実務では遭遇しないシチュエーションだと思うのでここら辺の細かい事情はレビューの対象外として良さそうですね

- `initialDate`あたりの処理、`YYYY-MM-DD`の形式でフォーマットしているかと思いますが、dayjsなどのライブラリ使うと１発なので、機会があれば使ってみてくださいませ〜〜

- planListとcompleteListに振り分ける処理、map使ってもいいかなーと思いましたが、mapを使うと2回ループを回す必要があるので1回のforループ内で処理してしまうのが処理効率的には良さそうですね💡
ループの計算量を無視できる程度のデータ量ならmapの方がわかりやすそうですが、ケースバイケースなところありそうだなぁと思いましたっ！

**Comment by mizushima1226** [comment]
https://github.com/generationokd/raisetech_todo/blob/7c7492b85b7b1f040cc3fa0fca8633866d7ffe9c/src/hooks/useInitialData.ts#L8-L11

Reactではデータを持つときはステートで保持するのが通常です💡
実装を見ていて「今回は練習としてAPIから取得したデータを初期値として使用する」のような意図を感じました（違うかもですが。。。）
ので、レビュー対象ではないかな？と思ってますが、
「データはステートで保持する」のところ認識だけ合っていれば大丈夫と思います👍

**Comment by mizushima1226** [comment]
おそらく実装漏れ？だと思いますが、ここのTS化はmustですねっ👀
useRecoilXXXのジェネリクスも省略できるかと！

https://github.com/generationokd/raisetech_todo/blob/7c7492b85b7b1f040cc3fa0fca8633866d7ffe9c/src/store/userState.js#L1-L5

**Comment by generationokd** [comment]
> ナイスバリデーションですっ
> `duration`、5000で統一してもいいかなぁと思いましたが、5000と2000で使い分けしているんですね💡
> どんなこだわりがあるのか気になります👀👀👀
> https://github.com/generationokd/raisetech_todo/blob/7c7492b85b7b1f040cc3fa0fca8633866d7ffe9c/src/pages/TodoList.tsx#L50-L57

最初は2000で統一していましたが、エラーのときはしっかり気づいてもらうために、少し長めの5000にしてみました。
改めて考えると、5000で統一で問題なさそうですね💡
完了時に長く表示されすぎるのも邪魔かと思いましたが、×ボタンで消せますし😀

**Comment by generationokd** [comment]
> 見た感じだとおそらく本当に必要な情報（提供したい情報）は`planList`と`completeList`でしょうか？👀
> それだと`apiData`はステートとして保持はせず、axios.getのレスポンスを受け取っているthen句の中で
> `planList`と`completeList`にデータを設定するのが良さそうです💡
> （実装の意図が伝わりやすくなる気がします💡）
> 
> https://github.com/generationokd/raisetech_todo/blob/7c7492b85b7b1f040cc3fa0fca8633866d7ffe9c/src/hooks/useInitialData.ts#L20-L27

はい！必要な情報は`planList`と`completeList`です。
確かに`apiData`は無駄なデータですね！全然意識できていなかったです。
ありがとうございます😀

**Comment by generationokd** [comment]
> https://github.com/generationokd/raisetech_todo/blob/7c7492b85b7b1f040cc3fa0fca8633866d7ffe9c/src/hooks/useInitialData.ts#L43-L70
> 
> * for文のiの設定、データが多いので10や20区切りで絞り込んでいる感じですかね👀
>   あくまでも今回は練習で、実務では遭遇しないシチュエーションだと思うのでここら辺の細かい事情はレビューの対象外として良さそうですね

はい！今回は練習用で、データの件数を減らすために、ユーザーID毎に10件ずつデータ取得できるようにしています。

> * `initialDate`あたりの処理、`YYYY-MM-DD`の形式でフォーマットしているかと思いますが、dayjsなどのライブラリ使うと１発なので、機会があれば使ってみてくださいませ〜〜

ありがとうございます！
dayjs等、見てみます👀

> * planListとcompleteListに振り分ける処理、map使ってもいいかなーと思いましたが、mapを使うと2回ループを回す必要があるので1回のforループ内で処理してしまうのが処理効率的には良さそうですね💡
>   ループの計算量を無視できる程度のデータ量ならmapの方がわかりやすそうですが、ケースバイケースなところありそうだなぁと思いましたっ！

今回は少し特殊な初期データセットをしたかったため、慣れているforを採用させていただきました💡

**Comment by generationokd** [comment]
> https://github.com/generationokd/raisetech_todo/blob/7c7492b85b7b1f040cc3fa0fca8633866d7ffe9c/src/hooks/useInitialData.ts#L8-L11
> 
> Reactではデータを持つときはステートで保持するのが通常です💡
> 実装を見ていて「今回は練習としてAPIから取得したデータを初期値として使用する」のような意図を感じました（違うかもですが。。。）
> ので、レビュー対象ではないかな？と思ってますが、
> 「データはステートで保持する」のところ認識だけ合っていれば大丈夫と思います👍

はい、おっしゃる通りこの部分は「今回は練習としてAPIから取得したデータを初期値として使用する」といった意図で実装しています。
「データはステートで保持する」の認識は問題ないです👍
